src/bin/pg_autoctl/primary_standby.c

```diff
@@ -984,3 +984,53 @@ check_postgresql_settings(LocalPostgresServer *postgres, bool *settings_are_ok)
 	pgsql_finish(pgsql);
 	return result;
 }
+
+
+/*
+ * primary_wait_until_standby_has_caught_up loops over a SQL query on the
+ * primary that checks the current reported LSN from the standby's replication
+ * slot.
+ */
+bool
+primary_wait_until_standby_has_caught_up(LocalPostgresServer *postgres)
+{
+	PGSQL *pgsql = &(postgres->sqlClient);
+
+	char standbyCurrentLSN[PG_LSN_MAXLENGTH] = { 0 };
+	bool reachedLSN = false;
+
+	log_info("Waiting until standby node has caught-up to LSN %s",
+			 postgres->currentLSN);
+
+	/* ensure some WAL level traffic to move things forward */
+	if (!pgsql_checkpoint(pgsql))
+	{
+		log_error("Failed to checkpoint after promotion");
+		return false;
+	}
+
+	/* loop until we have a slot that went past the current LSN */
+	while (!reachedLSN)
+	{
+		if (!pgsql_one_slot_has_reached_target_lsn(pgsql,
+												   postgres->currentLSN,
+												   standbyCurrentLSN,
+												   &reachedLSN))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		if (!reachedLSN)
+		{
+			log_info("Standby reached LSN %s, waiting for LSN %s",
+					 standbyCurrentLSN, postgres->currentLSN);
+			pg_usleep(AWAIT_PROMOTION_SLEEP_TIME_MS * 1000);
+		}
+	}
```

Comment:
Good point. The local currentLSN is updated at every keeper loop when we fetch the local Postgres metadata on the primary, so it's recent enough on each loop. Will see about adding a (short) timeout in that loop and allow the main FSM loop logic to kick-in.

Reply:
> so it's recent enough on each loop

What I meant is that if on the first try:
Enable sync replication
LSN is 100 after enabling sync replication

Secondary LSN is 50 
Primary goes back to FSM

On second try:
Enable sync replication (no-op)
currentLSN is now 150

Secondary LSN is now 120

At that point we secondary has caught up enough, but because we updated `currentLSN` it's we do not notice. So again we go back to the FSM.

So it would be nice if we save the LSN right after we enable sync replication and save that throughout multiple FSM loops (and unset it when we enable async replication or something).

Reply:
I see. Yes, indeed, that's how it should be done.

Reply:
Done now.